engine/src/main/java/io/camunda/zeebe/streamprocessor/records/RecordBatch.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.streamprocessor.records;
+
+import io.camunda.zeebe.protocol.record.RecordType;
+import io.camunda.zeebe.protocol.record.RejectionType;
+import io.camunda.zeebe.protocol.record.ValueType;
+import io.camunda.zeebe.protocol.record.intent.Intent;
+import io.camunda.zeebe.util.buffer.BufferWriter;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Spliterator;
+import java.util.function.Consumer;
+
+public final class RecordBatch implements ModifiableRecordBatch {
+
+  final List<UnmodifiableRecordBatchEntry> recordBatchEntries = new ArrayList<>();
+  private int batchSize;
+  private final int maxBatchSize;
+
+  public RecordBatch(final int maxBatchSize) {
+    this.maxBatchSize = maxBatchSize;
+  }
+
+  @Override
+  public void appendRecord(
+      final long key,
+      final int sourceIndex,
+      final RecordType recordType,
+      final Intent intent,
+      final RejectionType rejectionType,
+      final String rejectionReason,
+      final ValueType valueType,
+      final BufferWriter valueWriter) {
+    final var recordBatchEntry =
+        RecordBatchEntry.createRecordBatchEntry(
+            key,
+            sourceIndex,
+            recordType,
+            intent,
+            rejectionType,
+            rejectionReason,
+            valueType,
+            valueWriter);
+    final var entryLength = recordBatchEntry.getLength();
+
+    if (batchSize + entryLength >= maxBatchSize) {
```

Comment:
:question: How is the max batch size calculated? Because we had this problem with the dispatcher, where even if you said `4MB`, it wasn't really true because of alignment and double-framing, so we had to use a function (which delegates all way down to the dispatcher :disappointed:) to compute that on every entry.

Reply:
Right now it is set as 4 MB, but I also thought to give it either less or adjust the dispatcher size.